### PR TITLE
signin & signup added

### PR DESCRIPTION
**Pull request by anees-rehman-brainx**
**MERN Todo app**

1. database integrated
2. User SignIn added
3. User SignUP added

### CODE REVIEW

File: routes/todo-route.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const express = require('express');
+const router = express.Router();
+
+const {createTodo, getAllTodos,getTodoByUserId,getTodoByTodoId, deleteTodo,deleteTodoByUserId, updateTodo} = require('../controller/todo-controller');
```
Comment (muhammad-zeeshan-brainx) [inline routes/todo-route.js]: Please export all functions in object and then import that object in this route file.

**Review by muhammad-zeeshan-brainx — CHANGES_REQUESTED** [review]

File: controller/todo-controller.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
+
+const todoService = require("../service/todo-service")
+
+
+// Create Todo
+const createTodo = async (req, res) => {
+    try {
+        const {message} = req.body;
+        const userId = req.params.userId;
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js:9]: get user id from token not the params

File: controller/todo-controller.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        const date = new Date();
+
+        const todo ={
+            userId : userId,
+            todoId : Math.floor(Math.random() * 100) + userId,
+            message : message,
+            createdAt : date
+        }
+
+        try {
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js:19]: use only single try catch block

File: controller/todo-controller.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    } catch (error) {
+        res.status(404).json(error)
+    }
+}
+
+// get all todos
+const getAllTodos = async (req, res) => {
+
+    try {
+        const todos = todoService.getAllTodos();
```
Comment (ali-iqbal-brainx) [inline controller/todo-controller.js:36]: use await for the async calls

File: controller/user-controller.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+const userService = require('../service/user-service');
+const tokenService = require('../service/token-service')
+
+
+// function to register user
+const register = async(req, res) => {
+
+    const {username, email, password, confirmPassword} = req.body;
+
+    if(username ==  "" || email === "" || password === "" || confirmPassword === ""){
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:14]: use proper condition for error handling(undefined case will not be handle by this cond)

File: controller/user-controller.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    }
+   
+    const isEmailExist =await userService.validateEmail(email);
+
+    if(isEmailExist){
+        res.status(404).json("Email already exists");
+        return;
+    }
+
+    if(password != confirmPassword){
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:26]: use !== instead !=

File: controller/user-controller.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    const salt = await bcrypt.genSalt(10);
+    const hashedPassword = await bcrypt.hash(password, salt);
+    
+    const user = {
+        username : username,
+        email : email,
+        password : hashedPassword
+    }
+
+    const savedUser =await userService.addUser(user);
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:41]: use try catch block for the error that are unhandled

File: controller/user-controller.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        const {email, password} = req.body;
+
+        if(email === "" || password === ""){
+            res.status(400).json("All feilds are required..");
+            return;
+        }
+
+        const existedUser = await userService.getUserByEmail(email);
+
+        if(!existedUser){
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:69]: use chain operator and length operator

File: controller/user-controller.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            return;
+        }
+        else{
+            const payload = {
+                _id : existedUser[0]._id,
+                username : existedUser[0].username,
+                email : existedUser[0].email
+            }
+          
+            const secretKey = process.env.JWT_SECRET + password;
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:87]: not needed to add pass in secret key

File: controller/user-controller.js [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            const token = jwt.sign(payload, secretKey, {expiresIn : '30min'});
+
+            try {
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:90]: avoid using nested try catch

File: controller/user-controller.js [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        return;
+    }
+
+    if(isUserExist.password != oldPassword){
+        res.status(404).json("Incorrect old password..");
+        return;
+    }
+
+    if(newPassword != confirmNewPassword){
+        res.status(404).json("New Password does't match");
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:131]: use proper status code for all errors

File: controller/user-controller.js [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        return;
+    }
+
+    if(newPassword == oldPassword){
+        res.status(404).json("Old password and new password must be different");
+        return;
+    }
+
+    userService.updatePassword(isUserExist.userId, newPassword);
+    res.status(200).json("Password updated successfully....");
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:141]: return response only once

File: controller/user-controller.js [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+    const user = userService.getUserByEmail(email);
+
+    if(!user){
+        res.status(404).json("Email does't exist");
+        return;
+    }
+
+    // creating a unique token for password reset link
+    const secretKey = process.env.JWT_SECRET + user.password;
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:157]: user is an array password does not exist

File: controller/user-controller.js [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    const payload = {email : user.email, id : user.id};
+    const token = jwt.sign(payload, secretKey, {expiresIn : '15m'});
+
+    // Password resett link that is to be mailed to user 
+    const link = `http://localhost:${process.env.port}/user/reset-password/${user.userId}/${token}`;
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:162]: put host in env files instead of only port no.

File: controller/user-controller.js [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+
+}
+
+
+// Function to resett password
+const resetPassword = async(req, res) => {
+    try {
+        const {userId, token} = req.params;
+
+        const isUserExist = users.find(user => user.userId === userId);
```
Comment (ali-iqbal-brainx) [inline controller/user-controller.js:175]: use await for async calls

File: modals/todo-modal.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        type : String,
+        required : true
+    },
+    message : {
+        type : String,
+        required : true
+    },
+    todoStatus : {
+        type : String,
+        enum : ["incomplete", "inProgress", "completed"],
```
Comment (ali-iqbal-brainx) [inline modals/todo-modal.js:19]: put constants in a separate file

File: modals/user-modal.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+        required : true, 
+        minLength : 10, 
+        lowercase : true
+    },
+    password : {
+        type : String, 
+        required : true, 
+        minLength : 8, 
+    },
+    createdAt : {
```
Comment (ali-iqbal-brainx) [inline modals/user-modal.js:20]: use timestamp instead creating you own fields

File: service/user-service.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+// get all users
+const getAllUsers =async () => {
+    const users = await userModal.find();
+    return users;
+}
+
+
+// get user by ID
+const getUserById = (id) => {
+    const existedUser = users.find(user => user.userId === id);
```
Comment (ali-iqbal-brainx) [inline service/user-service.js:34]: user.find is not a function

File: service/user-service.js [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    return existedUser ? existedUser : false;
+}
+
+
+// get user by email
+const getUserByEmail = async (email) => {
+    return userModal.find({email : email});
```
Comment (ali-iqbal-brainx) [inline service/user-service.js:41]: use await for async calls

File: service/user-service.js [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+
+// change username
+const updateUsername = (userId, username) => {
+    const index = users.findIndex(user => user.userId === userId);
```
Comment (ali-iqbal-brainx) [inline service/user-service.js:47]: error

**Review by ali-iqbal-brainx — CHANGES_REQUESTED** [review]
-use only single try catch block(3)
-get user id from token not the params(2)
-use await for the async calls(5)
-use proper condition for error handling(undefined case will not be handle by this cond)(3)
-use !== instead !=(2)
-use try catch block for the error that are unhandled
-use chain operator and length operator
-not needed to add pass in secret key
-use proper status code for all errors
-return response only once
-user is an array password does not exist
-put host in env files instead of only port no
-use timestamp instead creating you own fields
-user.find is not a function(2)

**Pull request merged**